features/430-passwordless/implementation.md

```diff
@@ -0,0 +1,40 @@
+# Implementation Guide
+
+## Schema
+
+The provider tables (i.e. identities) will be kept, but password hash in
+password identity will be split out as a new type of authenticator (password
+authenticator).
+
+Existings authenticators for MFA would be kept and compatible with the new
+design. Each authenticator would be associated with a user (i.e. 1 user to
+many authenticators).
+
+## Authentication Session
+
+Authentication sessions would have four kind of steps:
+- Select identity: `identity:select`
+    - A list of usable identities type would be available for selection.
+    - User must provide sufficient information to determine the identity
+      (e.g. login ID), to proceed.
+- Perform primary authentication: `identity:authn`
+    - A list of usable primary authenticator types would be available for
+      selection.
+    - Some of the available authenticator may have not been set up for the
+      identity. These authenticators should always fail the authentication.
+- Perform secondary authentication: `mfa:authn`
+    - A list of usable secondary authenticators (set up for the user) would be
+      available for selection.
+    - At this step, we have sufficient confidence about user's identity to list
+      usable authenticators of the user.
+- Setup secondary authentication: `mfa:setup`
+    - User must setup one secondary authenticator to proceed.
+
+For Auth API, client will always recieve authentication session after
+`identity:authn` step. i.e. `mfa:authn` or `mfa:setup`.
+
+## API models
+
+Previously, we have two types of identities: `password` and `oauth`. Under the
+new design, `password` type should be renamed to `login_id`. However, for
```

Comment:
I think we should better rename this. Let's do a search to see if any project depend on the value of `password`.

Reply:
I think we used it in `Identity` object and the gateway header? Maybe we can just ask if anyone is using it, if no then we change it?🙈

Reply:
Also for the current headers, we had authenticator. We will need to update those headers too to support list?

Reply:
It is possible to maintain full backward compatibility, by keeping the 'password' identity type and fill in authenticator headers using secondary authenticator only. The question is do we want to keep it that way.